### ISSUE TITLE
Better adaptation configuration interface

### ISSUE THREAD

**Issue by xukai92** [TuringLang/Turing.jl#491]
Related:
https://github.com/TuringLang/Turing.jl/issues/448
https://github.com/TuringLang/Turing.jl/pull/485#issuecomment-418275625

**Comment by yebai** [comment]
Another related issue: #447

**Comment by yebai** [comment]
@mohamed82008 can you give this a try?

**Comment by trappmartin** [comment]
This might also be related #482

**Comment by yebai** [comment]
Perhaps @xukai92 can post a set of design goals for this issue.

**Comment by mohamed82008** [comment]
The current signature of `sample` is:
```julia
function sample(model::Function, alg::T;
                                chunk_size=CHUNKSIZE,               # set temporary chunk size
                                save_state=false,                   # flag for state saving
                                resume_from=nothing,                # chain to continue
                                reuse_spl_n=0,                      # flag for spl re-using
                                adapt_conf=STAN_DEFAULT_ADAPT_CONF  # adapt configuration
                               ) where T<:Hamiltonian
```
from https://github.com/TuringLang/Turing.jl/blob/7cb84a5a12dd604fae6b4b0e983b499d4dabb5ac/src/samplers/hmc.jl#L87.

The parameter `adapt_conf` is passed as the second argument to `Sampler` so that's the warmup sample size according to https://github.com/TuringLang/Turing.jl/issues/448#issuecomment-418088533. This is already available as a kwarg but the name may be obscure.

Other `sample` functions are:

https://github.com/TuringLang/Turing.jl/blob/166c2fba15c3debf0739cf763d3838d7d08b2533/src/samplers/ipmcmc.jl#L95

https://github.com/TuringLang/Turing.jl/blob/7cb84a5a12dd604fae6b4b0e983b499d4dabb5ac/src/samplers/gibbs.jl#L52

https://github.com/TuringLang/Turing.jl/blob/7cb84a5a12dd604fae6b4b0e983b499d4dabb5ac/src/samplers/is.jl#L38

https://github.com/TuringLang/Turing.jl/blob/7cb84a5a12dd604fae6b4b0e983b499d4dabb5ac/src/samplers/mh.jl#L111

https://github.com/TuringLang/Turing.jl/blob/7cb84a5a12dd604fae6b4b0e983b499d4dabb5ac/src/samplers/pgibbs.jl#L82

https://github.com/TuringLang/Turing.jl/blob/7cb84a5a12dd604fae6b4b0e983b499d4dabb5ac/src/samplers/pmmh.jl#L104

https://github.com/TuringLang/Turing.jl/blob/7cb84a5a12dd604fae6b4b0e983b499d4dabb5ac/src/samplers/smc.jl#L72

https://github.com/TuringLang/Turing.jl/blob/7cb84a5a12dd604fae6b4b0e983b499d4dabb5ac/src/samplers/support/stan-interface.jl#L15

With the exception of the last one, all take `model` and `alg` as the first 2 parameters. But not all have the same kwargs. I will see how I can improve the interface. Hints are welcome!

**Comment by yebai** [comment]
Hi @mohamed82008, thanks for looking into this. The main goals of this issue are

- Implement our own `Adapt` and `Hmc` related types. Turing currently re-uses the ones from `Stan.jl`, which can be found [here](https://github.com/goedman/Stan.jl/blob/master/src/types/sampletype.jl)

- Cleanup the `WarnUpManager` type. Perhaps merge with `Adapt`

- Cleanup the `sample` function so that they have consistent keyward arguments.

**Comment by yebai** [comment]
Closed by #793
